connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java

```diff
@@ -0,0 +1,220 @@
+        this.time = time;
+
+        Map<String, String> metricsTags = new LinkedHashMap<>(); // currently empty
+        MetricConfig metricConfig = new MetricConfig().samples(config.getInt(CommonClientConfigs.METRICS_NUM_SAMPLES_CONFIG))
+                                            .timeWindow(config.getLong(CommonClientConfigs.METRICS_SAMPLE_WINDOW_MS_CONFIG), TimeUnit.MILLISECONDS)
+                                            .recordLevel(Sensor.RecordingLevel.forName(config.getString(CommonClientConfigs.METRICS_RECORDING_LEVEL_CONFIG)))
+                                            .tags(metricsTags);
+        List<MetricsReporter> reporters = config.getConfiguredInstances(CommonClientConfigs.METRIC_REPORTER_CLASSES_CONFIG, MetricsReporter.class);
+        reporters.add(new JmxReporter(JMX_PREFIX));
+        this.metrics = new Metrics(metricConfig, reporters, time);
+        log.debug("Registering Connect metrics with JMX for worker '{}'", workerId);
+        AppInfoParser.registerAppInfo(JMX_PREFIX, workerId);
+    }
+
+    /**
+     * Get the worker identifier.
+     * @return the worker ID; never null
+     */
+    public String workerId() {
+        return workerId;
+    }
+
+    /**
+     * Get the {@link Metrics Kafka Metrics} that are managed by this object and that should be used to
+     * add sensors and individual metrics.
+     *
+     * @return the Kafka Metrics instance; never null
+     */
+    public Metrics metrics() {
+        return metrics;
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name.
+     *
+     * @param groupName the name of the metric group; may not be null
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName) {
+        return group(groupName, false);
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name and the given tags.
+     *
+     * @param groupName    the name of the metric group; may not be null
+     * @param tagKeyValues pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, String... tagKeyValues) {
+        return group(groupName, false, tagKeyValues);
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name and the given tags.
+     *
+     * @param groupName       the name of the metric group; may not be null
+     * @param includeWorkerId true if the tags should include the worker ID
+     * @param tagKeyValues    pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
```

Comment:
Agree. BTW, can't wait until we can start using `computeIfAbsent()`, which would make this even easier:

    public MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
        return groupsByName.computeIfAbsent(groupName, key->{
            Map<String, String> tags = tags(includeWorkerId ? workerId : null, tagKeyValues);
            return new MetricGroup(key, tags);
        });
    }

Thanks for the suggestion.

Reply:
You bet.

Reply:
Sorry, I'm not sure I understood the point of agreement. `computeIfAbsent` will still need a `ConcurrentMap`. I was saying that `synchronized` is going to do more harm than good here. And the use of concurrent map is practically fine. You agree with that, or leaving it the way it is?

Reply:
Yes, `ConcurrentMap` is better and I've changed the PR to reflect this. My point with the `computeIfAbsent()` is that it will be able to be even simpler and more efficient with JDK 8 (the `MetricGroup` is only created if it is needed). Until then, I went with your suggestion.

Reply:
Oh, I just noticed. Then `synchronized` is not needed anymore.

Reply:
It was removed from the other versions of `group` but not from here.

Reply:
Gotcha. It's fixed now.